engine/src/main/java/org/terasology/rendering/dag/nodes/CopyToVRFrameBuffersNode.java

```diff
@@ -0,0 +1,118 @@
+import org.terasology.rendering.world.WorldRenderer.RenderingStage;
+import static org.terasology.rendering.opengl.OpenGLUtils.renderFullscreenQuad;
+import static org.terasology.rendering.opengl.ScalingFactors.FULL_SCALE;
+
+public class CopyToVRFrameBuffersNode extends ConditionDependentNode {
+    private static final ResourceUrn LEFT_EYE_FBO = new ResourceUrn("engine:leftEye");
+    private static final ResourceUrn RIGHT_EYE_FBO = new ResourceUrn("engine:rightEye");
+    // TODO: make these configurable options
+
+    @In
+    private OpenVRProvider vrProvider;
+
+    @In
+    private WorldRenderer worldRenderer;
+
+    @In
+    private Config config;
+
+    @In
+    private ScreenGrabber screenGrabber;
+
+    @In
+    private DisplayResolutionDependentFBOs displayResolutionDependentFBOs;
+
+    private RenderingConfig renderingConfig;
+    private Material finalPost;
+    private FBO leftEye;
+    private FBO rightEye;
+
+    @Override
+    public void initialise() {
+        renderingConfig = config.getRendering();
+        finalPost = worldRenderer.getMaterial("engine:prog.post");
+        requiresCondition(() -> renderingConfig.isVrSupport());
+        leftEye = requiresFBO(new FBOConfig(LEFT_EYE_FBO, FULL_SCALE,
+                FBO.Type.DEFAULT).useDepthBuffer(),displayResolutionDependentFBOs);
+        rightEye = requiresFBO(new FBOConfig(RIGHT_EYE_FBO,FULL_SCALE,
+                FBO.Type.DEFAULT).useDepthBuffer(),displayResolutionDependentFBOs);
+        if (this.vrProvider != null) {
+            logger.info("Left eye FBOID:" + Integer.toString(leftEye.fboId));
+            logger.info("Right eye FBOID:" + Integer.toString(leftEye.fboId));
+            logger.info("OpenVR init done.");
+            vrProvider.texType[0].handle = leftEye.colorBufferTextureId;
+            vrProvider.texType[0].eColorSpace = JOpenVRLibrary.EColorSpace.EColorSpace_ColorSpace_Gamma;
+            vrProvider.texType[0].eType = JOpenVRLibrary.EGraphicsAPIConvention.EGraphicsAPIConvention_API_OpenGL;
+            vrProvider.texType[0].write();
+            vrProvider.texType[1].handle = rightEye.colorBufferTextureId;
+            vrProvider.texType[1].eColorSpace = JOpenVRLibrary.EColorSpace.EColorSpace_ColorSpace_Gamma;
+            vrProvider.texType[1].eType = JOpenVRLibrary.EGraphicsAPIConvention.EGraphicsAPIConvention_API_OpenGL;
+            vrProvider.texType[1].write();
+        }
+    }
+
+    @Override
+    public void process() {
+        PerformanceMonitor.startActivity("rendering/copyToVRFrameBuffers");
+        logger.info("Process");
+        finalPost.enable();
+        if (!renderingConfig.isVrSupport()) {
+            logger.warn("CopyToVRFrameBufferNode processed in non-VR mode.");
+        } else {
+            renderFinalStereoImage(worldRenderer.getCurrentRenderStage());
+        }
+        PerformanceMonitor.endActivity();
+    }
+
+    private void renderFinalStereoImage(RenderingStage renderingStage) {
```

Comment:
Not quite there yet but moving in the right direction. ;)

The goal with FBOs is that a node requires one on initialization and that FBO is automagically bound before the process() method is executed. This is achieved just by calling requireFBO() like you did, no need for further .bind() calls on the FBO.

The problem is this node requires two FBOs, one for each eye. In the future much of the current nodelist will be duplicated into parallel sets of nodes when VR is enabled, so this class would be instantiated twice and each instance would just require one FBO. For the time being however you need to have the right FBO bound at the right time.

(....after a 10 mins pensive pause...)
(...started writing something... stopped again...)

Ok, I think we need to take a step back.

What does this node do? For every frame it first uses the finalPost material to render the image for the left eye into the device's FBO. Then the renderer does a second pass through all nodes, reaches this node and the image for the right eye is generated, again into the device's FBO. Then the frame is submitted, which I seem to understand is when the hardware/driver for the HMD takes over the content of the two FBOs and actually displays the image in the headset.

In both rendering stages however, the FinalPostProcessingNode is processed too. So if I'm reading everything correctly, this last node renders a mono image to display, twice per frame.

It seems to me these aspects are related and should be tackled together. It's all about rendering and displaying the right thing on the right device (HMD vs monitor). This is different from how we used to do things for Oculus: OpenVR seems to require two separate (undistorted) color buffers when Oculus needed a single pre-composed and pre-distorted image.

So. I think we need to refactor the two last nodes in the list, CopyToVRFrameBuffersNode and FinalPostProcessingNode, and make three nodes out of them, with narrower, more precise responsibilities.

1. First we need a FinalPostProcessing node whose _sole_ responsibility is to render the final image into the FINAL FBO, whatever the circumstances (mono/left/right).
2. Then we need a CopyImageToHMD node which (if VR is enabled) takes the content of the final FBO and _copies it_ (not renders it) into the appropriate device-bound FBO. When the second eye is processed this node is also responsible for copying the content of both left/right FBOs back onto FINAL, so that it contains a split screen with the images for both eyes.
3. Finally, a CopyImageToScreen node _copies_ the content of the FINAL FBO to the default FBO, the display, but only when in MONO mode or when the second eye is processed. And this node should also include the functionality to handle screenshots, which is currently in FinalPostProcessingNode.

Notice the emphasis on the _copies_ verbs: the last rendering to a quad happens in the FinalPostProcessing node. The following nodes use inexpensive copy operations instead. 

How does this sound? Once we have these three nodes up and running we can revisit the issue of binding the right FBO for the appropriate eye at the right time.